matrix-api.c

```diff
@@ -431,31 +434,42 @@ static gchar *_build_request(PurpleAccount *acct, const gchar *url,
     if(body != NULL)
         g_string_append(request_str, body);
 
+    if(extra_data != NULL)
+        g_string_append_len(request_str, extra_data, extra_len);
+
+    *total_len = request_str->len; /* TODO: Surely there must be a function!? */
     return g_string_free(request_str, FALSE);
 }
 
 
 /**
  * Start an HTTP call to the API
  *
- * @param method  HTTP method (eg "GET")
- * @param body    body of request, or NULL if none
- * @param max_len maximum number of bytes to return from the request. -1 for
- *                default (512K).
+ * @param method      HTTP method (eg "GET")
+ * @param extra_headers  Extra HTTP headers to add
+ * @param body        body of request, or NULL if none
+ * @param extra_data  raw binary data to be sent after the body
+ * @param extra_len   The length of the raw binary data
+ * @param max_len     maximum number of bytes to return from the request. -1 for
+ *                    default (512K).
  *
  * @returns handle for the request, or NULL if the request couldn't be started
  *   (eg, invalid hostname). In this case, the error_callback will have
  *   been called already.
  */
 static MatrixApiRequestData *matrix_api_start(const gchar *url,
```

Comment:
done; we now have matrix_api_start_full with the extras from these two patches and have kept the matrix_api_start as the wrapper.